lib/server_manager.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/fatih/color"
+	"golang.zx2c4.com/wireguard/wgctrl"
+	"golang.zx2c4.com/wireguard/wgctrl/wgtypes"
+)
+
+type ServerInfo struct {
+	i      uint16
+	cancel context.CancelFunc
+}
+
+// ServerManager handles creating and terminating servers on ips
+type ServerManager struct {
+	wgClient      *wgctrl.Client
+	ipt           *iptables.IPTables
+	key           wgtypes.Key
+	password      string
+	ctx           context.Context
+	waitGroup     *sync.WaitGroup
+	wgBlock       netip.Prefix
+	wgBlockPerIp  uint
+	activeServers map[netip.Addr]ServerInfo
+
+	// freeIndices and nextFreeIndex together track usage of the range 0..numWgBocks
+	freeIndices   []uint16 // stack of indices that are free
+	nextFreeIndex uint16   // next free index not in the stack
+}
+
+// NewServerManager creates a new server manager
+func NewServerManager(wgBlock netip.Prefix, wgBlockPerIp uint, ctx context.Context, key wgtypes.Key, password string) (*ServerManager, error) {
+	// Make a shared WireGuard client.
+	wgClient, err := wgctrl.New()
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize wgctrl: %v", err)
+	}
+
+	ipt, err := iptables.New(iptables.IPFamily(iptables.ProtocolIPv4), iptables.Timeout(5))
+	if err != nil {
+		return nil, fmt.Errorf("failed to initialize iptables: %v", err)
+	}
+
+	// Display the public key, just for information.
+	fmt.Printf("%s %s\n",
+		color.New(color.Bold).Sprint("server public key:"),
+		key.PublicKey().String())
+
+	sm := new(ServerManager)
+	sm.wgClient = wgClient
+	sm.ipt = ipt
+	sm.key = key
+	sm.password = password
+	sm.ctx = ctx
+	sm.waitGroup = new(sync.WaitGroup)
+	sm.wgBlock = wgBlock.Masked()
+	sm.wgBlockPerIp = wgBlockPerIp
+	sm.activeServers = make(map[netip.Addr]ServerInfo)
+	return sm, nil
+}
+
+// allocateIndex attempts to allocate a new index for a server
+func (sm *ServerManager) allocateIndex() (uint16, error) {
+	l := len(sm.freeIndices)
+	if l > 0 {
+		defer func() {
+			sm.freeIndices = sm.freeIndices[:l-1]
+		}()
+		return sm.freeIndices[l-1], nil
+	} else {
+		wgBlockCount := uint16(1) << (sm.wgBlockPerIp - uint(sm.wgBlock.Bits()))
+		if sm.nextFreeIndex == wgBlockCount {
+			return 0xFFFF, fmt.Errorf("no more free indices for provided number of wireguard blocks")
+		}
+		defer func() {
+			sm.nextFreeIndex += 1
+		}()
+		return sm.nextFreeIndex, nil
+	}
+}
+
+// freeIndex frees the specified index
+func (sm *ServerManager) freeIndex(i uint16) {
+	sm.freeIndices = append(sm.freeIndices, i)
+}
+
+// Start creates a new server on the specified ip.
+func (sm *ServerManager) Start(ip netip.Addr) error {
+	i, err := sm.allocateIndex()
+	if err != nil {
+		return err
+	}
+	subctx, cancel := context.WithCancel(sm.ctx)
+
+	srv := &Server{
+		Key:      sm.key,
+		BindAddr: ip,
+		Password: sm.password,
+		Index:    i,
+		Ipt:      sm.ipt,
+		WgClient: sm.wgClient,
+		WgCidr:   netip.PrefixFrom(AfterCountIpBlock(sm.wgBlock.Addr().Next(), sm.wgBlockPerIp, uint(i)+1), int(sm.wgBlockPerIp)),
```

Comment:
Might be misreading here, but did you mean to take `i` instead of `i+1`? Like

```go
subnetStart := AfterCountIpBlock(sm.wgBlock.Addr(), sm.wgBlockPerIp, i)
netip.PrefixFrom(subnetStart.Next(), int(sm.wgBlockPerIp))
```